pinot-query-planner/src/test/java/org/apache/pinot/query/planner/logical/StagesTestBase.java

```diff
@@ -0,0 +1,230 @@
+   */
+  public SimpleChildBuilder<MailboxReceiveNode> exchange(
+      int nextStageId, SimpleChildBuilder<? extends PlanNode> childBuilder) {
+    return (stageId, mySchema, myHints) -> {
+      PlanNode input = childBuilder.build(stageId);
+      MailboxSendNode mailboxSendNode = new MailboxSendNode(nextStageId, null, List.of(input), stageId, null, null,
+          null, false, null, false);
+      MailboxSendNode old = _stageRoots.put(nextStageId, mailboxSendNode);
+      Preconditions.checkState(old == null, "Mailbox already exists for stageId: %s", nextStageId);
+      return new MailboxReceiveNode(stageId, null, List.of(), nextStageId, null, null, null, null,
+          false, false, mailboxSendNode);
+    };
+  }
+
+  /**
+   * Creates a table scan node with the given table name.
+   */
+  public SimpleChildBuilder<TableScanNode> tableScan(String tableName) {
+    return (stageId, mySchema, myHints) -> new TableScanNode(stageId, mySchema, myHints, List.of(), tableName,
+        Collections.emptyList());
+  }
+
+  /**
+   * Looks for the mailbox that corresponds to the given stageId.
+   * @throws IllegalStateException if the mailbox is not found.
+   */
+  public MailboxSendNode stage(int stageId) {
+    MailboxSendNode result = _stageRoots.get(stageId);
+    Preconditions.checkState(result != null, "Mailbox not found for stageId: %s", stageId);
+    return result;
+  }
+
+  /**
+   * Returns a builder that can be used to create a new mailbox send node.
+   *
+   * Whenever this builder is used to create a node, the created node will be added to the list of mailboxes.
+   *
+   * It is usually recommended to use {@link #exchange(int, SimpleChildBuilder)} instead of this method, given that
+   * {@code exchange} creates a pair of send and receive mailboxes and deals with the stageId management.
+   */
+  public SimpleChildBuilder<MailboxSendNode> sendMailbox(
+      int newStageId, SimpleChildBuilder<? extends PlanNode> childBuilder) {
+    return (stageId, mySchema, myHints) -> {
+      PlanNode input = childBuilder.build(stageId);
+      MailboxSendNode mailboxSendNode = new MailboxSendNode(newStageId, mySchema, List.of(input), stageId, null, null,
+          null, false, null, false);
+      MailboxSendNode old = _stageRoots.put(stageId, mailboxSendNode);
+      Preconditions.checkState(old == null, "Mailbox already exists for stageId: %s", stageId);
+      return mailboxSendNode;
+    };
+  }
+
+  /**
+   * A builder that can be used to create a child node.
+   *
+   * It is not expected for test writers to implement this class. Instead it is recommended to use methods like
+   * {@link #exchange(int, SimpleChildBuilder)}, {@link #join(SimpleChildBuilder, SimpleChildBuilder)},
+   * {@link #tableScan(String)} and others to chain instances of this class.
+   */
+  public interface ChildBuilder<P extends PlanNode> {
+    P build(int stageId, @Nullable DataSchema dataSchema, @Nullable PlanNode.NodeHint hints);
```

Comment:
This isn't currently being used anywhere directly? All external uses seem to be through `SimpleChildBuilder`?

Reply:
No, it is not used externally and probably should never be. But we need it to create the machinary that runs withDataSchema and withHints